src/vs/workbench/contrib/mcp/common/discovery/installedMcpServersDiscovery.ts

```diff
@@ -0,0 +1,151 @@
+		@IMcpRegistry private readonly mcpRegistry: IMcpRegistry,
+		@ITextModelService private readonly textModelService: ITextModelService,
+	) {
+		super();
+	}
+
+	public start(): void {
+		const throttler = this._register(new Throttler());
+		this._register(this.mcpWorkbenchService.onChange(() => throttler.queue(() => this.sync())));
+		this.sync();
+	}
+
+	private async getServerIdMapping(resource: URI, pathToServers: string[]): Promise<Map<string, Location>> {
+		const store = new DisposableStore();
+		try {
+			const ref = await this.textModelService.createModelReference(resource);
+			store.add(ref);
+			const serverIdMapping = getMcpServerMapping({ model: ref.object.textEditorModel, pathToServers });
+			return serverIdMapping;
+		} catch {
+			return new Map();
+		} finally {
+			store.dispose();
+		}
+	}
+
+	private async sync(): Promise<void> {
+		try {
+			const collections = new Map<string, [IMcpConfigPath | undefined, McpServerDefinition[]]>();
+			const mcpConfigPathInfos = new ResourceMap<Promise<IMcpConfigPath & { locations: Map<string, Location> } | undefined>>();
+			for (const server of this.mcpWorkbenchService.local) {
+				if (!server.local) {
+					continue;
+				}
+
+				let mcpConfigPathPromise = mcpConfigPathInfos.get(server.local.mcpResource);
+				if (!mcpConfigPathPromise) {
+					mcpConfigPathPromise = (async (local: ILocalMcpServer) => {
+						const mcpConfigPath = this.mcpWorkbenchService.getMcpConfigPath(local);
+						const locations = mcpConfigPath?.uri ? await this.getServerIdMapping(mcpConfigPath?.uri, mcpConfigPath.section ? [...mcpConfigPath.section, 'servers'] : ['servers']) : new Map();
+						return mcpConfigPath ? { ...mcpConfigPath, locations } : undefined;
+					})(server.local);
+					mcpConfigPathInfos.set(server.local.mcpResource, mcpConfigPathPromise);
+				}
+
+				const config = server.local.config;
+				const mcpConfigPath = await mcpConfigPathPromise;
+				const collectionId = `mcp.config.${mcpConfigPath ? mcpConfigPath.id : 'unknown'}`;
+
+				let definitions = collections.get(collectionId);
+				if (!definitions) {
+					definitions = [mcpConfigPath, []];
+					collections.set(collectionId, definitions);
+				}
+
+				definitions[1].push({
+					id: `${collectionId}.${server.local.name}`,
+					label: server.local.name,
+					launch: config.type === 'http' ? {
+						type: McpServerTransportType.HTTP,
+						uri: URI.parse(config.url),
+						headers: Object.entries(config.headers || {}),
+					} : {
+						type: McpServerTransportType.Stdio,
+						command: config.command,
+						args: config.args || [],
+						env: config.env || {},
+						envFile: config.envFile,
+						cwd: config.cwd
+							// if the cwd is defined in a workspace folder but not absolute (and not
+							// a variable or tilde-expansion) then resolve it in the workspace folder
+							? (!isAbsolute(config.cwd) && !config.cwd.startsWith('~') && !config.cwd.startsWith('${') && mcpConfigPath?.workspaceFolder
```

Comment:
This is missing some changes around the cwd that I made to fix a candidate last week (see L123 in the deleted configMcpDiscovery.ts). I plan to go back and clean them up this iteration, but we should pull that in to avoid regression.